allenact/embodiedai/mapping/mapping_models/active_neural_slam.py

```diff
@@ -0,0 +1,646 @@
+
+    # First we place the egocentric map view into the center
+    # of a map that has the same size as the allocentric map
+
+    nbatch, c, ego_h, ego_w = map_probs_egocentric.shape
+    allo_h, allo_w = allocentric_map_height_width
+
+    max_view_range = math.sqrt((ego_w / 2.0) ** 2 + ego_h ** 2)
+    if min(allo_h, allo_w) / 2.0 < max_view_range:
+        raise NotImplementedError(
+            f"The shape of your egocentric view (ego_h, ego_w)==({ego_h, ego_w})"
+            f" is too large relative the size of the allocentric map (allo_h, allo_w)==({allo_h}, {allo_w})."
+            f" The height/width of your allocentric map should be at least {2 * max_view_range} to allow"
+            f" for no information to be lost when rotating the egocentric map."
+        )
+
+    full_size_ego_map_update_probs = map_probs_egocentric.new(
+        nbatch, c, *allocentric_map_height_width
+    ).fill_(0)
+
+    assert (ego_h % 2, ego_w % 2, allo_h % 2, allo_w % 2) == (
+        0,
+    ) * 4, "All map heights/widths should be divisible by 2."
+
+    x1 = allo_w // 2 - ego_w // 2
+    x2 = x1 + ego_w
+    z1 = allo_h // 2
+    z2 = z1 + ego_h
+    full_size_ego_map_update_probs[:, :, z1:z2, x1:x2] = map_probs_egocentric
+
+    # Now we'll rotate and translate `full_size_ego_map_update_probs`
+    # so that the egocentric map view is positioned where it should be
+    # in the allocentric coordinate frame
+
+    # To do this we first need to rescale our allocentric xz coordinates
+    # so that the center of the map is (0,0) and the top left corner is (-1, -1)
+    # as this is what's expected by the `affine_grid` function below.
+    rescaled_xzrs_allocentric = xzrs_allocentric.clone().detach().float()
+    rescaled_xzrs_allocentric[:, :2] *= (
+        100.0 / resolution_in_cm
+    )  # Put x / z into map units rather than meters
+    rescaled_xzrs_allocentric[:, 0] /= allo_w / 2  # x corresponds to columns
+    rescaled_xzrs_allocentric[:, 1] /= allo_h / 2  # z corresponds to rows
+    rescaled_xzrs_allocentric[:, :2] -= 1.0  # Re-center
+
+    x = rescaled_xzrs_allocentric[:, 0]
+    z = rescaled_xzrs_allocentric[:, 1]
+    theta = (
+        -rescaled_xzrs_allocentric[:, 2] * DEGREES_TO_RADIANS
+    )  # Notice the negative sign
+
+    cos_theta = theta.cos()
+    sin_theta = theta.sin()
+    zeroes = torch.zeros_like(cos_theta)
+    ones = torch.ones_like(cos_theta)
+
+    theta11 = torch.stack([cos_theta, -sin_theta, zeroes], 1)
+    theta12 = torch.stack([sin_theta, cos_theta, zeroes], 1)
+    theta1 = torch.stack([theta11, theta12], 1)
+
+    theta21 = torch.stack([ones, zeroes, x], 1)
+    theta22 = torch.stack([zeroes, ones, z], 1)
+    theta2 = torch.stack([theta21, theta22], 1)
+
+    grid_size = (nbatch, c, allo_h, allo_w)
+    rot_grid = F.affine_grid(theta1, grid_size)
+    trans_grid = F.affine_grid(theta2, grid_size)
+
+    return F.grid_sample(
```

Comment:
this is really awesome (didn't know there was affine_grid/grid_sample in torch)

Reply:
Yah, that they are able to GPU accelerate this operation is pretty amazing, I'm very curious how they manage this.